Source/DafnyCore/Compilers/Rust/Dafny-compiler-rust-proofs.dfy

```diff
@@ -0,0 +1,362 @@
+
+
+
+
+
+include "../Dafny/AST.dfy"
+
+module {:extern "DCOMPProofs"} {:compile false} DCOMPProofs refines DCOMP {
+
+  ghost predicate IsDafnyIdTail(s: string) {
+    |s| == 0 ||
+    (s[0] in "aqkhd_?'#." && IsDafnyIdTail(s[1..]))
+  }
+
+  ghost predicate IsDafnyId(s: string) {
+    && |s| > 0 && s[0] in "aqkhd" &&
+       IsDafnyIdTail(s[1..])
+  }
+  lemma IsDafnyIdImpliesDafnyIdTail(s: string)
+    requires IsDafnyId(s)
+    ensures IsDafnyIdTail(s)
+  {}
+
+  function dafnyEscape(s: string): string {
+    if |s| == 0 then "" else
+    if s[0] == '_' then "__" + dafnyEscape(s[1..]) else
+    if s[0] == '?' then "_q" + dafnyEscape(s[1..]) else
+    if s[0] == '\'' then "_k" + dafnyEscape(s[1..]) else
+    if s[0] == '#' then "_h" + dafnyEscape(s[1..]) else
+    [s[0]] + dafnyEscape(s[1..])
+  }
+
+  function dafnyEscapeReverse(s: string): string {
```

Comment:
`dafnyUnescape` would be a more familiar name IMHO

Reply:
Renamed to `DafnyUnescape` in the next PR